### PR TITLE
[feat] Add support for the LSF scheduler and LLNL's `lrun` parallel launcher

### LINKED ISSUE #1055: Add support for the LSF scheduler
I got a couple of requests to implement support for the LSF scheduler. We could do a basic implementation as with PBS that we could test with the Slurm LSF wrappers.

### PR DESCRIPTION
**Pull request by rngoodner**
This adds basic support for LSF and lrun. LSF is a scheduler and lrun is a launcher used on the Lawrence Livermore National Laboratory Sierra Systems. I have used this implementation to run tests on the Lassen cluster successfully.

I followed the sge scheduler implementation that worked off the pbs implementation as an example.

Closes #1055

### CODE REVIEW

**Comment by teojgo** [comment]
Ok to test

File: reframe/core/launchers/mpi.py [hunk 1]
```diff
@@ -124,3 +124,26 @@ def command(self, job):
             ret.append(opt)
 
         return ret
+
+
+@register_launcher('lrun')
+class LrunLauncher(JobLauncher):
+    '''LLNL's custom parallel job launcher.'''
+    def command(self, job):
+        num_tasks = job.num_tasks or 1
```
Comment (vkarak) [inline reframe/core/launchers/mpi.py:133]: I'm not sure that this is needed. `job.num_tasks` is always > 0. If it is set to zero, then the scheduler will try to do a flexible allocation if supported.

File: reframe/core/launchers/mpi.py [hunk 2]
```diff
@@ -124,3 +124,26 @@ def command(self, job):
...
+        num_tasks_per_node = job.num_tasks_per_node or 1
+        num_nodes = int(num_tasks // num_tasks_per_node)
```
Comment (vkarak) [inline reframe/core/launchers/mpi.py:135]: The `//` is by definition the integer division and both operands are guaranteed to be integers:

```suggestion
        num_nodes = num_tasks // num_tasks_per_node
```

File: reframe/core/launchers/mpi.py [hunk 3]
```diff
@@ -124,3 +124,26 @@ def command(self, job):
...
+@register_launcher('lrun-gpu')
+class LrungpuLauncher(JobLauncher):
+    '''LLNL's custom parallel job launcher w/ CUDA aware Spectum MPI.'''
+    def command(self, job):
+        num_tasks = job.num_tasks or 1
+        num_tasks_per_node = job.num_tasks_per_node or 1
+        num_nodes = int(num_tasks // num_tasks_per_node)
+        return ['lrun', '-M "-gpu"',
+                '-N', str(num_nodes),
+                '-T', str(num_tasks_per_node)]
```
Comment (vkarak) [inline reframe/core/launchers/mpi.py:149]: You could possibly inherit from `LrunLauncher()` and implement `command()` as follows:

```suggestion
    def command(self, job):
        return super().command(job) + ['-M "-gpu"']
```

File: reframe/core/schedulers/lsf.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+@register_scheduler('lsf')
+class LsfJobScheduler(PbsJobScheduler):
+    def __init__(self):
+        self._prefix = '#BSUB'
+        self._submit_timeout = rt.runtime().get_option(
+            f'schedulers/@{self.registered_name}/job_submit_timeout'
+        )
+
+    def emit_preamble(self, job):
+        num_tasks = job.num_tasks or 1
```
Comment (vkarak) [inline reframe/core/schedulers/lsf.py:34]: See my comment above:

```suggestion
        num_tasks = job.num_tasks
```

File: reframe/core/schedulers/lsf.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+        if not jobs:
+            return
+
+        completed = osext.run_command(
+            f'bjobs -noheader {" ".join(job.jobid for job in jobs)}')
+        if completed.returncode != 0:
+            raise JobSchedulerError(
+                f'bjobs failed with exit code {completed.returncode} '
+                f'(standard error follows):\n{completed.stderr}'
```
Comment (vkarak) [inline reframe/core/schedulers/lsf.py:89]: This should be enough:

```suggestion
        completed = _run_strict(
            f'bjobs -noheader {" ".join(job.jobid for job in jobs)}'
        )
```

File: reframe/core/schedulers/lsf.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            job_regex = (r'(?P<jobid>\d+)\s+'
+                         r'(?P<user>\S+)\s+'
+                         r'(?P<status>\S+)\s+'
+                         r'(?P<queue>\S+)')
+            job_match = re.search(job_regex, line)
+            if job_match:
+                job_status[job_match['jobid']] = job_match['status']
+
+        for job in jobs:
+            # job id not found
```
Comment (vkarak) [inline reframe/core/schedulers/lsf.py:105]: Move these comments inside the corresponding `if`s.

File: reframe/core/schedulers/lsf.py [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            if job.jobid not in job_status:
+                self.log(f'Job {job.jobid} not known to scheduler, '
+                         f'assuming job completed')
+                job._state = 'COMPLETED'
+                job._completed = True
+
+            # job done
+            elif job_status[job.jobid] in ['DONE', 'EXIT']:
```
Comment (vkarak) [inline reframe/core/schedulers/lsf.py:113]:
```suggestion
            elif job_status[job.jobid] in ('DONE', 'EXIT'):
```

**Review by vkarak — CHANGES_REQUESTED** [review]
Thanks a lot @rngoodner for your PR! It looks very good overall. I have only some minor comments regarding the implementation. We would also need the corresponding entries about the added scheduler and parallel launchers in the docs.

**Comment by vkarak** [comment]
@AnibalMG could you test if this LSF backend works for your case?

File: reframe/core/launchers/mpi.py (hunk 1 above)
Comment (rngoodner) [reply reframe/core/launchers/mpi.py]: done

File: reframe/core/launchers/mpi.py (hunk 2 above)
Comment (rngoodner) [reply reframe/core/launchers/mpi.py]: done

File: reframe/core/launchers/mpi.py (hunk 3 above)
Comment (rngoodner) [reply reframe/core/launchers/mpi.py]: done

File: reframe/core/schedulers/lsf.py (hunk 5 above)
Comment (rngoodner) [reply reframe/core/schedulers/lsf.py]: done

File: reframe/core/schedulers/lsf.py (hunk 7 above)
Comment (rngoodner) [reply reframe/core/schedulers/lsf.py]: done

File: reframe/core/schedulers/lsf.py (hunk 4 above)
Comment (rngoodner) [reply reframe/core/schedulers/lsf.py]: done

**Comment by rngoodner** [comment]
@vkarak I have addressed all of your comments and also updated the docs. I have also re-tested on the LLNL Lassen cluster and everything is still working for my needs.

**Review by vkarak — APPROVED** [review]
Lgtm! I just did a couple of minor style fixes.

**Pull request merged**